### ISSUE TITLE
Understand how elasticsearch snapshots works

### ISSUE THREAD

**Issue by deepthidevaki** [camunda/zeebe#9104]
The goal is to understand how we can use elasticsearch snapshots for backups of all components in camunda 8 that uses elasticsearch for storage.

* How to take elasticsearch snapshots
* What are the gurantees related to failures? - master not available, node restarts etc.
* What are the consistency guarantees in a snapshot - when there are multiple indices/shards?

**Comment by deepthidevaki** [comment]
Handling/detecting failures:

Snapshot process can fail, but we can monitor the status of a snapshot using the snapshot api.
Monitor status of a snapshot using https://www.elastic.co/guide/en/elasticsearch/reference/current/get-snapshot-api.html    
![image](https://user-images.githubusercontent.com/1997478/162997172-628c2304-7b36-4342-88fb-f0efc66d9f10.png)


- Can we retake a failed snapshot with the same Id?
I think it is not possible. The snapshotId must be unique. If a snapshot failed, the failed snapshot is still trackable. That means a snapshot with that id exists, although it is a failed one. But we can delete a snapshot. So probably we can delete a failed snapshot and take a new snapshot with the same id. Should verify this.
It is important to identify a snapshot so that we can determine which snapshot belong to the global backup involving multiple components.

**Comment by deepthidevaki** [comment]
Consistency guarantees in a snapshot:

https://www.elastic.co/guide/en/elasticsearch/reference/current/snapshot-restore.html#how-snapshots-work
> To back up an index, a snapshot makes a copy of the index’s segments and stores them in the snapshot repository. Since segments are immutable, the snapshot only needs to copy any new segments created since the repository’s last snapshot.


> A snapshot doesn’t represent a cluster at a precise point in time. Instead, each snapshot includes a start and end time. The snapshot represents a view of each shard’s data at some point between these two times.


https://www.linkedin.com/pulse/how-elasticsearch-snapshots-work-steve-mushero
> Snapshots are instantaneous snaps of the state when the snapshot STARTS, so any data indexed after the snapshot start will NOT be included (get it next time). This is because the first step on each node is building the segment list, and by definition, any 'new' data after that point will be in new segments that aren't included in the list.


When a snapshot process is started, it collects the current segments and then start copying it to the backup storage. Any newly written data will be in new segments so won't be in the backup. 

There is no information about how backups for multiple indices are coordinated. It is highly probable that there is no coordination, and the backups for different indices starts at different time. So if we write X to index-1 and Y to index-2, it is possible that Y is in the backup but X is not. Whether this is a problem when taking backups depends on how Operate/Optimize uses elasticsearch.

**Comment by deepthidevaki** [comment]
### TLDR: 
* There is no guarantee on what will be included in a backup in each shards and each indices.
* This means there are possible data loss when we take ElasticSearch snapshots while still writing to it. We should consider this when defining the solutions for backups of each application that uses ES. As examples:
    - No guarantee between indices can cause [these issues in Optimize ](https://github.com/camunda/zeebe/issues/9106#issuecomment-1097684586) 
    - No guarantee between shards can result in missing zeebe records in the backup, which might break assumptions in the applications consuming them. That is backup may contain records with position 9 and 11, but record 10 is missing.

----

After looking into ES snapshotting code and reading [docs](https://javadoc.io/doc/org.elasticsearch/elasticsearch/latest/org/elasticsearch/snapshots/package-summary.html), here is what I understood about elasticsearch's snapshotting process.


Here is a simplified process of taking snapshots in ES.
![image](https://user-images.githubusercontent.com/1997478/163364793-0c48e2b2-9d54-47f9-9c24-2f1517714474.png)

All writes happened before "Collect segments" will be included in the backup, and everything after that won't be included.

Now, consider there are two sequential writes that happened in parallel to the snapshotting process
![image](https://user-images.githubusercontent.com/1997478/163364825-a5df3488-a42d-4ea5-a666-a0bf3da132c1.png)

The operations happen in the following order.
1. Start taking backup
2. Start taking backup of shard 1
3. Write record 10 -> Goes to shard 1
4. write record 11 -> Goes to shard 2
5. Start taking bakcup of shard 2
6. Finished taking backup

In this scenario the backup will contain record 11, but not record 10, even though record 10 was written before record 11. That means there is no guarantee between the backups of multiple shards from the same index.

With this information the following statements from the docs makes more sense.
> A snapshot doesn’t represent a cluster at a precise point in time. Instead, each snapshot includes a start and end time. The snapshot represents a view of each shard’s data at some point between these two times.

Shards are like partitions in zeebe (in some way). There are no coordination between the partitions. Writes to different shards are uncoordinated. And When taking snapshots each shard is taking the backup independently.

Replace shards with index in the above scenario, and we get the same result.

This makes sense as ES is not a relational database. So consistency guarantees will be more relaxed. Also worth noting that [the consistency guarantees defined in ES is per document.](https://www.elastic.co/guide/en/elasticsearch/guide/2.x/optimistic-concurrency-control.html). It is not per index or per shard, or no global guarantees.

**Comment by deepthidevaki** [comment]
We now have enough information about elasticsearch snapshotting process and its guarantees. With this information we can find a solution that work around the lack of guarantees. 

Closing this issue. But feel free to add comments/questions.

**Comment by deepthidevaki** [comment]
@npepinpe brought to my attention about the refresh intervals in ES. After looking into it, it looks like this may also affect what is included in the Snapshot. https://www.elastic.co/guide/en/elasticsearch/guide/2.x/translog.html

When we "put" a document into ES, it is first written to an in-memory buffer and an on-disk "translog" (Writing to translog guarantees that the writes are not lost after a restart.). Then ES periodically refreshes in which a new segment is created and the in-memory buffer is written to the segment. When we take a snapshot, it collects the segments to be backed-up. So it is possible that whatever is in the in-memory buffer is not included in the snapshot. It is not clear to me whether there is an auto-refresh before a snapshot is taken. If we want to ensure that all "writes" before the snapshot operation is included in the snapshot, we may have to do a force refresh using the Refresh api.
